### ISSUE TITLE
Next alpha release for MIT folks

### ISSUE THREAD

**Issue by yarikoptic** [dandi/s3invsync#81]
Could we plan on what else to do to cut the next release for folks to try out? E.g. should following still open ATM issues be addressed first 

- #53
- #54

?

**Comment by jwodder** [comment]
- For #53, it depends on what exactly's causing the crash.  If it's just running out of memory normally, then #64 should fix that, but if it's some side effect of building with musl, then I don't know.

- For #54, I've added more logging in order to help diagnose the problem.  Other than doing more test runs, I'm not sure what to do for that.

**Comment by yarikoptic** [comment]
ok, I closed #53 .  @jwodder please do the needed run to check for #54 on drogon (not enough space on typhon I think ATM) with desired level of verbosity etc to track down/resolve that issue.

**Comment by yarikoptic** [comment]
FWIW, I freed up typhon space enough to run it there too now

**Comment by jwodder** [comment]
@yarikoptic I believe #83 fixes #54.  I ran the code on drogon with `--path-filter '^zarr/aa'`, and it successfully ran for over ten hours before failing due to an error on S3's end.

**Comment by yarikoptic** [comment]
Great, thanks!

Re error on S3 end -- I think all kinds of 500s could happen. Do we have retries in place for whenever interactions with S3 fail for some 5xx reason?

**Comment by yarikoptic** [comment]
Great, thanks!

Re error on S3 end -- I think all kinds of 500s could happen. Do we have retries in place for whenever interactions with S3 fail for some 5xx reason?

**Comment by jwodder** [comment]
@yarikoptic No.  I'm not even sure how to discern 5xx from other errors with this SDK.

**Comment by yarikoptic** [comment]
What particular SDK? (sounds "too bad to be true")

**Comment by yarikoptic** [comment]
I see them testing retries on 500 for some components e.g.
https://github.com/awslabs/aws-sdk-rust/blob/7791b008320b70ea59880079ba901908f3712f5c/sdk/aws-config/src/imds/client.rs#L809
and issues like
- https://github.com/awslabs/aws-sdk-rust/issues/1163

hinting that in general it should be retried.  What specific error did you receive?   might need an issue to be reported against SDK.

**Comment by jwodder** [comment]
@yarikoptic The error was:

```
Error: failed to get object at s3://dandiarchive/zarr/aae64bf8-594e-4c7b-a1a2-9bcbc8638186/0/0/0/4/8/156?versionId=o_N.RxYr0iL73ZNug020YtzrSaWoJce2

Caused by:
    0: service error
    1: unhandled error (InternalError)
    2: Error { code: "InternalError", message: "We encountered an internal error. Please try again.", aws_request_id: "XB0VWDXJH3351P6T", s3_extended_request_id: "QdlprriTEB4HT9v6CzktdTBJUZf8XFecTws0XGaTVEJ6rjRukIxT2Pu+/NdMBOq8fNRzsAHGTk9cK3GmDAFjvXxREvGj0+rcb6P/AEpNYkY=" }
```

**Comment by yarikoptic** [comment]
ah, so it looks very similar to that
 
- https://github.com/awslabs/aws-sdk-rust/issues/1163

and IMHO should be retried for this GET object. Could you please report an issue?  meanwhile -- wouldn't it be possible to configure retries on any `InternalError`s?

**Comment by jwodder** [comment]
@yarikoptic [The SDK uses a default retry config](https://docs.aws.amazon.com/sdk-for-rust/latest/dg/retries.html), so it's possible it did retry but it just didn't show up in the logs due to a low log level.  I seem to recall from working on dandidav that the AWS SDK logs a lot of noise at DEBUG and TRACE level, so increasing the logs might not be such a good idea.

**Comment by yarikoptic** [comment]
do you know if it logs at INFO or WARNING level when initial fail before retry happens?  I failed to track it down in the code there

**Comment by jwodder** [comment]
@yarikoptic So what, if anything, should we do about retrying?

**Comment by yarikoptic** [comment]
- if you think that it didn't retry at all -- we better submit an issue against SDK
- if you think that it did retry but not enough, and I think that is what we could do anyways to robustify, let's change default retrying policy to retry up to e.g. 10 times with exponential backoff like they describe in https://docs.aws.amazon.com/sdk-for-rust/latest/dg/retries.html .
  - if we encounter similar problem after introducing this change to retry policy, it would almost definitely mean it didn't retry and we must report.

**Comment by jwodder** [comment]
@yarikoptic I did not encounter the problem while performing the runs listed in #106.  Can we consider the problem resolved now and release v0.1.0-alpha.2?

**Comment by yarikoptic** [comment]
yes, please! and lets close this issue with the release

**Comment by jwodder** [comment]
@yarikoptic Release made: https://github.com/dandi/s3invsync/releases/tag/v0.1.0-alpha.2
